lnwallet/rpcwallet/remote_signer_client.go

```diff
@@ -0,0 +1,977 @@
+// enables an outbound remote signer.
+func NewOutboundClient(walletServer walletrpc.WalletKitServer,
+	signerServer signrpc.SignerServer,
+	streamFeeder SignCoordinatorStreamFeeder,
+	requestTimeout time.Duration) (*OutboundClient, error) {
+
+	if walletServer == nil || signerServer == nil {
+		return nil, errors.New("sub-servers cannot be nil when using " +
+			"an outbound remote signer")
+	}
+
+	if streamFeeder == nil {
+		return nil, errors.New("streamFeeder cannot be nil")
+	}
+
+	return &OutboundClient{
+		walletServer:    walletServer,
+		signerServer:    signerServer,
+		streamFeeder:    streamFeeder,
+		requestTimeout:  requestTimeout,
+		quit:            make(chan struct{}),
+		retryTimeout:    defaultRetryTimeout,
+		maxRetryTimeout: defaultMaxRetryTimeout,
+	}, nil
+}
+
+// Start starts the remote signer client. The function will continuously try to
+// setup a connection to the configured watch-only node, and retry to connect if
+// the connection fails until we Stop the remote signer client.
+func (r *OutboundClient) Start() error {
+	r.wg.Add(1)
+
+	// We'll continuously try setup a connection to the watch-only node, and
+	// retry to connect if the connection fails until we Stop the remote
+	// signer client.
+	go func() {
+		defer r.wg.Done()
+
+		for {
+			err := r.run()
+			if err != nil {
+				log.Errorf("Remote signer client error: %v",
+					err)
+			}
+
+			select {
+			case <-r.quit:
+				return
+			default:
+				log.Infof("Will retry to connect to "+
+					"watch-only node in: %v",
+					r.retryTimeout)
+
+				// Backoff before retrying to connect to the
+				// watch-only node.
+				select {
+				case <-r.quit:
+					return
+				case <-time.After(r.retryTimeout):
+				}
+			}
+
+			log.Infof("Retrying to connect to watch-only node")
+
+			// Increase the retry timeout by 50% for every retry.
+			r.retryTimeout = time.Duration(float64(r.retryTimeout) *
+				retryMultiplier)
+
+			// But cap the retryTimeout at r.maxRetryTimeout
+			if r.retryTimeout > r.maxRetryTimeout {
+				r.retryTimeout = r.maxRetryTimeout
+			}
+		}
+	}()
+
+	return nil
+}
+
+// Stop stops the remote signer client.
+func (r *OutboundClient) Stop() error {
+	if r.stopped.Swap(true) {
+		return errors.New("remote signer client is already shut down")
+	}
+
+	log.Info("Remote signer client shutting down")
+
+	// Ensure that no new Run goroutines can start when we've initiated
+	// the stopping of the remote signer client.
+	r.wgMu.Lock()
+
+	if r.streamFeeder != nil {
+		r.streamFeeder.Stop()
+	}
+
+	close(r.quit)
+
+	r.wgMu.Unlock()
+
+	r.wg.Wait()
+
+	log.Debugf("Remote signer client shut down")
+
+	return nil
+}
+
+// run creates a new stream to the watch-only node, and starts processing and
+// responding to the sign requests that are sent over the stream. The function
+// will continuously run until the remote signer client is either stopped or
+// the stream errors.
+func (r *OutboundClient) run() error {
```

Comment:
i think this and the handshake methods can be simplified quite a lot just using context correctly:

```
func (r *OutboundClient) run(ctx context.Context) error {
	ctx, cancel := CancelOrQuit(ctx, r.quit)
	ctx, _ = context.WithTimeout(ctx, r.requestTimeout)

	// Cancel the stream context whenever we return from this function.
	defer cancel()

	log.Infof("Attempting to setup the watch-only node connection")

	// Try to get a new stream to the watch-only node.
	stream, streamCleanUp, err := r.streamFeeder.GetStream(ctx)
	if err != nil {
		return err
	}

	defer streamCleanUp()

	// Once the stream has been created, we'll need to perform the handshake
	// process with the watch-only node, before it will start sending us
	// requests.
	err = r.handshake(stream)
	if err != nil {
		return err
	}

	log.Infof("Completed setup connection to watch-only node")

	// Reset the retry timeout after a successful connection.
	r.retryTimeout = defaultRetryTimeout

	return r.processSignRequests(ctx)
}

// handshake performs the handshake process with the watch-only node. As we are
// the initiator of the stream, we need to send the first message over the
// stream. The watch-only node will only proceed to sending us requests after
// the handshake has been completed.
func (r *OutboundClient) handshake(stream StreamClient) error {
	// completeType is a type alias for the registration complete type,
	// created to keep line length within 80 characters.
	type completeType = walletrpc.RegistrationResponse_RegistrationComplete

	// TODO(viktor): This could be extended to include info about the
	// version of the remote signer in the future.
	// The RegistrationChallenge should also be set to a randomized string.
	signReg := &walletrpc.SignerRegistration{
		RegistrationChallenge: "registrationChallenge",
		RegistrationInfo:      "outboundSigner",
	}

	regType := &walletrpc.SignCoordinatorResponse_SignerRegistration{
		SignerRegistration: signReg,
	}

	registrationMsg := &walletrpc.SignCoordinatorResponse{
		RefRequestId:     handshakeRequestID,
		SignResponseType: regType,
	}

	err := stream.Send(registrationMsg)
	if err != nil {
		return err
	}

	// After the registration message has been sent, the signer node will
	// respond with a message indicating that it has accepted the signer
	// registration request if the registration was successful.
	msg, err := r.stream.Recv()
	if err != nil {
		return err
	}

	if msg.GetRequestId() != handshakeRequestID {
		return fmt.Errorf("initial response request id must "+
			"be %d, but is: %d", handshakeRequestID,
			msg.GetRequestId())
	}
	// Check the type of the response message.
	switch reqType := msg.GetSignRequestType().(type) {
	case *walletrpc.SignCoordinatorRequest_RegistrationResponse:
		switch rType := reqType.RegistrationResponse.
			GetRegistrationResponseType().(type) {
		// The registration was successful.
		case *completeType:
		// TODO(viktor): This should verify that the signature in the
		// complete message is valid.

		// An error occurred during the registration process.
		case *walletrpc.RegistrationResponse_RegistrationError:
			return fmt.Errorf("registration error: %s",
				rType.RegistrationError)

		}

	default:
		return fmt.Errorf("expected registration response, "+
			"but got: %T", reqType)
	}

	return nil
}
```

Reply:
Unfortunately this wouldn't fully work, but thanks a lot for the patch :(! I implemented as much of this as I could, but especially the `handshake` still needs to contain the `goroutines`.

The main reason that this wouldn't work is that we can't add a timeout to the context in the `Run` function, i.e.:
`ctx, _ = context.WithTimeout(ctx, r.requestTimeout)`

If we did, when that context times out, the stream would also be cancelled. That wouldn't work, as this stream is intended to be long lived.

As we can't add the timeout to the context, we need to ensure that we can timeout the `stream.Send` + `stream.Recv` in the handshake in another way. And therefore we need to have those running in a goroutine, which unfortunately creates some complexity in the `handshake` function.

Reply:
We could potentially add a `SendWithTimeout` & `ReceiveWithTimeout` function that moves the complexity and goroutine spawning one level up. If you think that it would be worth doing that. Let me know!

Reply:
Hmmm I've realised that it'd probably be cleaner to let the entire `handshake` function run in a goroutine that's spawn in the `run` function, and let the timeout be handled there, instead of launching the goroutines inside the handshake function. Then we could implement your suggested version of the `handshake` function :)!

Reply:
ok, posting full diff of my suggestion here. Note that this is a patch applied to `022ba399a14af86e32f35ac9ed9d96f882df85be`

[orrs.patch](https://github.com/user-attachments/files/18111378/orrs.patch)

Reply:
Niiiiiiiiiiiiiiiceeee 🚀🔥!!!!!!! Thanks a lot for the patch and suggestion @ellemouton! I totally agree that this simplifies the code, so thanks for taking so much time to suggest it :pray:!

I've implemented your suggested approach, with some slight modifications (mainly that I think it makes more sense that the `handshake` etc. defines their own timeout context, rather than leaving that to the caller). Have also closely checked that the behaviour for the remote signer client should be the same as the previous version :tada:!

Thanks again :pray: